event-svc/src/event/service.rs

```diff
@@ -187,47 +210,35 @@ impl EventService {
             }
         }
 
-        // Group events by their type
-        let mut valid_events = Vec::with_capacity(parsed_events.len());
-        let mut signed_events = Vec::with_capacity(parsed_events.len());
-        let mut time_events = Vec::with_capacity(parsed_events.len());
-        for event in parsed_events {
-            match event.event().as_ref() {
-                Event::Time(_) => {
-                    time_events.push(event);
-                }
-                Event::Signed(_) => {
-                    signed_events.push(event);
-                }
-                Event::Unsigned(_) => {
-                    // Unsigned events need no extra validation.
-                    valid_events.push(event);
-                }
-            }
-        }
-
-        let (
-            (valid_signed_events, invalid_signed_events),
-            (valid_time_events, invalid_time_events),
-        ) = try_join!(
-            Self::validate_signed_events(signed_events),
-            Self::validate_time_events(time_events)
-        )?;
+        let pending_to_insert = self.remove_pending(&parsed_events);
+        let to_validate = parsed_events.into_iter().chain(pending_to_insert).collect();
+
+        let (validated, pending) = if self.validate_events {
+            let ValidatedEvents {
+                valid,
+                pending,
+                invalid,
+            } = EventValidator::validate_events(&self.pool, source, to_validate).await?;
+            invalid_events.extend(invalid);
+            (valid, pending)
+        } else {
+            (
+                // just turn them into validated "for free"
+                to_validate.into_iter().map(|e| e.into()).collect(),
+                Vec::new(),
+            )
+        };
 
-        valid_events.extend(valid_signed_events);
-        valid_events.extend(valid_time_events);
-        invalid_events.extend(invalid_signed_events);
-        invalid_events.extend(invalid_time_events);
+        // we consider pending "invalid" to tell recon/api what happened.
+        // for recon, it's okay and we'll track them below and try to find events we need in the future,
+        // but if this is API (Immediate), we won't track them because they shouldn't exist yet.
+        pending.iter().for_each(|p| {
+            invalid_events.push(InvalidItem::RequiresHistory {
+                key: p.order_key().to_owned(),
+            })
+        });
 
-        Ok((valid_events, invalid_events))
-    }
-
-    pub(crate) async fn insert_events(
-        &self,
-        items: &[ReconItem<EventId>],
-        source: DeliverableRequirement,
-    ) -> Result<InsertResult> {
-        let (to_insert, mut invalid) = Self::validate_events(items).await?;
+        let to_insert: Vec<EventInsertable> = validated.into_iter().map(|e| e.into()).collect();
```

Comment:
yeah, probably makes sense.. I have been trying to avoid pass through functions but I think this has gotten big enough now. I didn't expect parsing to happen before originally and thought everything was going to be in the validator file.